File: rd-net/RdFramework/Impl/RdPerContextMap.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using JetBrains.Collections.Synchronized;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Serialization;

namespace JetBrains.Rd.Impl
{
    public class RdPerContextMap<K, V> : RdReactiveBase, IPerContextMap<K, V> where V : RdBindableBase where K: notnull
    {
        public RdContext<K> Context { get; }
        private readonly Func<Boolean, V> myValueFactory;
        private readonly IViewableMap<K, V> myMap;
        private ReactiveQueue<KeyValuePair<K, V>>? myQueue;

        public RdPerContextMap(RdContext<K> context, Func<bool, V> valueFactory)
        {
            myValueFactory = valueFactory;
            Context = context;
            myMap = new ViewableMap<K, V>(new SynchronizedDictionary<K, V>());
        }

        public bool IsMaster;

        public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader stream, IRdWireableDispatchHelper dispatchHelper)
        {
            // this entity does not receive messages
            var message = nameof(RdPerContextMap<K, V>) + " may not receive messages";
            Assertion.Fail(message);
        }

        protected override void PreInit(Lifetime lifetime, IProtocol proto)
        {
            base.PreInit(lifetime, proto);
            
            var queue = new ReactiveQueue<KeyValuePair<K, V>>();
            myQueue = queue;

            var protocolValueSet = proto.Contexts.GetValueSet(Context);
            proto.Scheduler.InvokeOrQueue(() =>
            {
              foreach (var k in myMap.Keys.Where(key => !protocolValueSet.Contains(key)).ToList()) 
                myMap.Remove(k);
            });
            
            protocolValueSet.View(lifetime, (contextValueLifetime, contextValue) =>
            {
              myMap.TryGetValue(contextValue, out var oldValue);

              var value = oldValue ?? myValueFactory(IsMaster);

              using (var cookie = contextValueLifetime.UsingExecuteIfAlive())
              {
                if (!cookie.Succeed)
                  return;
                
                value.WithId(proto.Identities.Mix(RdId, contextValue.ToString()));
                value.PreBind(contextValueLifetime, this, $"[{contextValue.ToString()}]");  
              }

              queue.Enqueue(new KeyValuePair<K, V>(contextValue, value));
            });
        }
        
        protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
        {
          base.Init(lifetime, proto, ctx);

          var queue = myQueue;
          if (queue == null)
          {
            if (proto.Lifetime.IsNotAlive)
              return;
            
            Assertion.Fail($"{nameof(myQueue)} must not be null");
          }

          queue.View(lifetime, pair =>
          {
            using var cookie = lifetime.UsingExecuteIfAlive();
            if (!cookie.Succeed)
              return;

            pair.Value.Bind();

            if (!myMap.ContainsKey(pair.Key))
            {
              proto.Scheduler.InvokeOrQueue(() =>
              {
                if (!myMap.ContainsKey(pair.Key))
                  myMap[pair.Key] = pair.Value;
              });
            }
          });
        }
        
        public V GetForCurrentContext()
        {
          var currentId = Context.ValueForPerContextEntity;
          if (Mode.IsAssertion) Assertion.AssertNotNull(currentId, "No value set for key {0}", Context.Key);
          if (TryGetValue(currentId, out var value))
            return value;
          Assertion.Fail("{0} has no value for Context {1} = {2}", Location, Context.Key, currentId);
          return default;
        }
        

        public void View(Lifetime lifetime, Action<Lifetime, KeyValuePair<K, V>> handler)
        {
          AssertThreading();
          myMap.View(lifetime, handler);
        }

        public void View(Lifetime lifetime, Action<Lifetime, K, V> handler)
        {
          AssertThreading();
          myMap.View(lifetime, handler);
        }

        public V this[K key]
        {
          get
          {
            if (!IsBound)
            {
              if (!myMap.ContainsKey(key))
                myMap[key] = myValueFactory(IsMaster);
            }

            AssertThreading();
            return myMap[key];
          }
        }

        public bool TryGetValue(K key, [MaybeNullWhen(false)] out V value)
        {
          if (!IsBound)
          {
            if (!myMap.ContainsKey(key))
              myMap[key] = myValueFactory(IsMaster);
          }
          
          AssertThreading();
          return myMap.TryGetValue(key, out value);
        }


        public static void Write(SerializationCtx context, UnsafeWriter writer, RdPerContextMap<K, V> value)
        {
            RdId.Write(writer, value.RdId);
        }

        public static RdPerContextMap<K, V> Read(SerializationCtx context, UnsafeReader reader, RdContext<K> key, Func<bool, V> func)
        {
            var id = RdId.Read(reader);
            return new RdPerContextMap<K, V>(key, func).WithId(id);
        }
        
        private class ReactiveQueue<T>
        {
          private Action<T>? myListener;
          private readonly Queue<T> myQueue = new();
          
          public void Enqueue(T value)
          {
            Action<T> action;
            lock (myQueue)
            {
              if (myListener is { } listener)
              {
                Assertion.Assert(myQueue.Count == 0);
                action = listener;
              }
              else
              {
                myQueue.Enqueue(value);
                return;
              }
            }

            action.Invoke(value);
          }

          public void View(Lifetime lifetime, Action<T> action)
          {
            while (lifetime.IsAlive)
            {
              T value;
              lock (myQueue)
              {
                if (myQueue.Count > 0)
                {
                  if (lifetime.IsNotAlive)
                    return;
                  
                  value = myQueue.Dequeue();
                }
                else
                {
                  lifetime.TryBracket(() =>
                  {
                    Assertion.Assert(myListener == null);
                    myListener = action;
                  }, () =>
                  {
                    lock (myQueue)
                      myListener = null;
                  });
                  
                  return;
                }
              }

              action(value);
            }
          } 
        }
    }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/RdPerContextMap.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.collections.SynchronizedMap
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.isAlive
import com.jetbrains.rd.util.lifetime.isNotAlive
import com.jetbrains.rd.util.reactive.AddRemove
import com.jetbrains.rd.util.reactive.IViewableMap
import com.jetbrains.rd.util.reactive.ViewableMap
import kotlin.collections.ArrayDeque

class RdPerContextMap<K: Any, V : RdBindableBase> private constructor(override val context: RdContext<K>, val valueFactory: (Boolean) -> V, private val internalMap: ViewableMap<K, V>) : RdReactiveBase(), IPerContextMap<K, V>, IViewableMap<K, V> by internalMap {
    constructor(context: RdContext<K>, valueFactory: (Boolean) -> V) : this(context, valueFactory, ViewableMap(SynchronizedMap()))

    override fun deepClone(): IRdBindable {
        return RdPerContextMap(context, valueFactory)
    }

    val changing = internalMap.changing
    var optimizeNested: Boolean = false

    private var queue: ReactiveQueue<Pair<K, V>>? = null

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        // this entity has no own messages
        error("RdPerContextMap at $location received a message" )
    }

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        val localQueue = ReactiveQueue<Pair<K, V>>()
        queue = localQueue


        val protocolValueSet = proto.contexts.getValueSet(context)
        proto.scheduler.invokeOrQueue {
            internalMap.keys.filter { !protocolValueSet.contains(it) }.forEach { internalMap.remove(it) }
        }

        protocolValueSet.view(lifetime) { contextValueLifetime, contextValue ->
            val oldValue = internalMap[contextValue]

            val value = (oldValue ?: valueFactory(master))

            contextValueLifetime.executeIfAlive {
                value.withId(proto.identity.mix(rdid, contextValue.toString()))
                value.preBind(contextValueLifetime, this, "[${contextValue}]")
            }

            localQueue.enqueue(Pair(contextValue, value))
        }
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        val localQueue = queue
        if (localQueue == null) {
            if (proto.lifetime.isNotAlive)
                return

            error("queue must not be null")
        }

        localQueue.view(lifetime) { pair ->
            lifetime.executeIfAlive {
                pair.second.bind()

                if (!internalMap.containsKey(pair.first)) {
                    proto.scheduler.invokeOrQueue {
                        if (!internalMap.containsKey(pair.first)) {
                            internalMap[pair.first] = pair.second
                        }
                    }
                }
            }
        }
    }

    override fun adviseAddRemove(lifetime: Lifetime, handler: (AddRemove, K, V) -> Unit) {
        assertThreading()
        internalMap.adviseAddRemove(lifetime, handler)
    }

    override fun view(lifetime: Lifetime, handler: (Lifetime, Map.Entry<K, V>) -> Unit) {
        assertThreading()
        internalMap.view(lifetime, handler)
    }

    override fun view(lifetime: Lifetime, handler: (Lifetime, K, V) -> Unit) {
        assertThreading()
        internalMap.view(lifetime, handler)
    }

    override operator fun get(key: K): V? {
        if(!isBound) {
            if (!internalMap.containsKey(key))
                internalMap[key] = valueFactory(false)
        }

        assertThreading()
        return internalMap[key]
    }

    override fun getForCurrentContext(): V {
        val currentId = context.valueForPerContextEntity ?: error("No ${context.key} set for getting value for it")
        return this[currentId] ?: error("No value in ${this.location} for ${context.key} = $currentId")
    }

    companion object {
        fun <K: Any, V : RdBindableBase> read(key: RdContext<K>, buffer: AbstractBuffer, valueFactory : (Boolean) -> V) : RdPerContextMap<K, V> {
            val id = buffer.readRdId()
            return RdPerContextMap(key, valueFactory).withId(id)
        }

        fun write(buffer: AbstractBuffer, map: RdPerContextMap<*, *>) {
            buffer.writeRdId(map.rdid)
        }
    }

    private class ReactiveQueue<T> {
        private var listener: ((T) -> Unit)? = null
        private val queue = ArrayDeque<T>()

        fun enqueue(value: T) {
            val action: (T) -> Unit
            synchronized(queue) {
                val localListener = listener
                if (localListener != null) {
                    assert(queue.size == 0)
                    action = localListener
                } else{
                    queue.add(value)
                    return
                }
            }

            action(value)
        }

        fun view(lifetime: Lifetime, action: (T) -> Unit) {
            while (lifetime.isAlive) {
                val value = synchronized(queue) {
                    if (queue.size > 0) {
                        if (lifetime.isNotAlive)
                            return

                        queue.removeFirst()
                    } else {
                        lifetime.bracketIfAlive({
                            assert(listener == null)
                            listener = action
                        }, {
                            synchronized(queue) {
                                listener = null
                            }
                        })

                        return
                    }
                }

                action(value)
            }
        }
    }
}